### PR TITLE
Refactor Winapp2.ini

### PR DESCRIPTION
**Pull request by MoscaDotTo**
Here I've started a refactor. Some comments:

Implemented broad use of nested wildcards and wildcards in detectfile paths

I merged together many entries. Many of these were situations like separate logs/cache cleaning, or situations were multiple similar entries are related. My rationale behind this is that it's a little exhausting having Skype installed and to be greeted with 5 extra checkboxes for components of Skype that otherwise come with no explanation. Things that a user might generally not want to clean alongside "junk" files (such as history, chat logs, etc) I did not merge

It's a pretty broad, sweeping change to a lot of parts of the file, so feel free to cherrypick/debate the usefulness of any changes. 

There are some pull requests pending that might not merge well with this also.

### CODE REVIEW

**Comment by ROCKNROLLKID** [comment]
I have informed @APMichael about this PR and will most likely close his. This will take a few days to review.

The first thing I noticed off the back is that you reorganized entries that I organized and put them out of ABC order, like for example, you put Aegisub Backups above Aegisub, when it should be below it for organization. Same thing with AVG entries and many others. IS there a reason why you put them back like this?

**Comment by MoscaDotTo** [comment]
I can undo the alphabetization change, it was automated by the (now quite aged) trim.bat, which likely picks up the space character as "before" the entry without the space character. I could also split this up into multiple pull requests to make it easier to review in "chunks" if that helps.

**Comment by MoscaDotTo** [comment]
I will see about updating WinappDebug to detect misalphabetized entry names (filekeys/regkeys will be a little more difficult but I will also see about that) to help with that problem in general.

**Comment by ROCKNROLLKID** [comment]
It might be better to separate these into chunks. That way we can merge some that are done, and hold off on others that need work.

I appreciate you updating Winappdebug. For now, we have been just relying on @az0 tool at BleachBit, but I, myself, don't know how to use it since it requires a bit of Python knowledge. His tool is very through as it is and I would like to maybe try and combine his tool with Winappdebug if it is possible. I talked to @ShaneGowland  in the past about this, but he said that Winappdebug currently doesn't check each line but rather the whole entry. Shane said this would need a re-write.

**Comment by MoscaDotTo** [comment]
Currently (our) WinappDebug reads the file line by line, which doesn't lend itself well to scanning things by entry, but doesn't make it impossible either. Since each entry is separated by a newline, we can detect each time we move to the next entry (this is how the numbering check works, it increments the number that it believes should be next to each Detect/File File/RegKey until it reaches a newline character, where it resets them to 1).

**Comment by ROCKNROLLKID** [comment]
Well you are doing good with it. I am not a programmer so I don't know how ti works internally, haha.

**Comment by MoscaDotTo** [comment]
I updated WinappDebug here: https://github.com/MoscaDotTo/Winapp2/commit/6a7cf648cb6b329f712426ce67d31aaa856d7d8e

It will tell you when things are alphabetical, but it comes with a slight quirk surrounding the "-" character which is that it ignores it, so entries like [X-Cleaner*] and [X-NetStat*] get sorted to XC and XN respectively instead of X- (which we currently place before other letters)

**Comment by ROCKNROLLKID** [comment]
Looks like it also ignores numbers, too. All in all, it looks good. We may have to add exceptions for certain false positive errors so people don't think they are real issues when they aren't.

And I guess Winappdebug can't count the number of entries because of what you said above?

**Comment by MoscaDotTo** [comment]
I updated WinappDebug again here: https://github.com/MoscaDotTo/Winapp2/commit/106ecba68d6c556e424f92dfd5c9dfafb9b1c99c

it now sorts things properly with a "-" character (by replacing it with two spaces for sorting purposes) and also counts the entries. It should also now properly detect entries that are beyond where they're meant to be alphabetically (Eg, [A...] in with the [B...]s) instead of just ignoring it. The trade-off for this is slightly less accurate ability to detect when entries are placed too early, or rather something of an inability to distinguish which entry in a pair is out of order between an out of place entry and its immediate sorted neighbor. 

Anyway! Please try this version and let me know if you like the results more. I haven't started on FileKey sorting yet, that may have to be a project for next week. :)

**Comment by CSGalloway** [comment]
How about modifying it so that it will check a user-specified INI file or
if not specified defaults to WinApp2.ini - that way other people can use it
on their custom rules

http://www.facebook.com/CSGalloway

**Comment by ROCKNROLLKID** [comment]
Just tested it and I think it is good. Thank you for doing this.

My only other request would be if it is possible to combine trim.bat, Winappparser and Winappdebug into 1 tool and maybe call it Winapp2analysis or something similar. Maybe when first load it up, it gives three options, like 1. Run Winappdebug, 2. Run Winappparser, and 3. Run Trim. How would you feel about having something like this?

Otherwise, I just wait for you to make the corrections to this PR and then I think we can merge it if no other issues appear.

**Comment by ROCKNROLLKID** [comment]
Don't know if you feel up to it, but if you could, could you apply the suggestions from upstream here: https://forum.piriform.com/topic/32310-winapp2ini-additions/?page=265

Everything from my post on the top and down needs to be done. Otherwise, I can do it after this gets merged myself.

**Comment by MoscaDotTo** [comment]
I'll update this PR soon after I finish correcting the many hundreds of remaining alphabetization fixes revealed by https://github.com/MoscaDotTo/Winapp2/commit/b7480c2a5b1fcab882cc3dae41b3ced7bc0ae00b

File: Winapp2.ini [hunk 1]
```diff
@@ -1559,8 +1535,7 @@ FileKey1=%AppData%\Adobe\Photoshop Album\*.*|Logse*.txt
 [Adobe Photoshop CS2 More*]
 LangSecRef=3021
 Detect=HKCU\Software\Adobe\Photoshop\9.0
-DetectFile1=%AppData%\Adobe\CameraRaw\Cache
-DetectFile2=%AppData%\Adobe\Bridge\Cache
+DetectFile1=%AppData%\Adobe\*\Cache
```
Comment (ROCKNROLLKID) [inline Winapp2.ini]: Do wildcards work in Detectfile?

File: Winapp2.ini [hunk 2]
```diff
@@ -1897,17 +1865,12 @@ FileKey2=%AppData%\aignes\website-watcher\config\log|*.*
...
-Default=False
-FileKey1=%LocalAppData%\AOL\AIM\cache|*.*|REMOVESELF
-
-[AIM Logs*]
+[AIM*]
 LangSecRef=3022
 DetectFile=%LocalAppData%\AOL\AIM
 Default=False
 FileKey1=%LocalAppData%\AOL\AIM|*.log|RECURSE
+FileKey2=%LocalAppData%\AOL\AIM\cache|*.*|REMOVESELF
```
Comment (ROCKNROLLKID) [inline Winapp2.ini]: We should leave cache files in there own entry as these files are meant to speed up a program and should only be removed if you need disk space. Also, most programs just regenerate cache files, which might make too many writes on SSDs.

File: Winapp2.ini [hunk 3]
```diff
@@ -2418,36 +2369,25 @@ FileKey1=%AppData%\Appupdater|appupdaterw.log
...
-Warning=This will delete your chat history.
 Detect1=HKCU\Software\MyPortal\AQQ
 Detect2=HKCU\Software\Wapster\AQQ
 Default=False
-FileKey1=%UserProfile%\WapSter\AQQ Folder\Profiles\*\Data\Temp|*.*
-FileKey2=%LocalAppData%\MyPortal\AQQ Folder\Profiles\*\Data\Temp|*.*
+FileKey1=%UserProfile%\WapSter\AQQ Folder\Profiles\*\Data\*Avatars|*.*
+FileKey2=%UserProfile%\WapSter\AQQ Folder\Profiles\*\Data\*Cache|*.*
+FileKey3=%LocalAppData%\MyPortal\AQQ Folder\Profiles\*\Data\*Avatars|*.*
+FileKey4=%LocalAppData%\MyPortal\AQQ Folder\Profiles\*\Data\*Cache|*.*
```
Comment (ROCKNROLLKID) [inline Winapp2.ini]: Same here. Please keep the cache files separate.

File: Winapp2.ini [hunk 4]
```diff
@@ -3241,7 +3120,7 @@ Section=Games
 Detect=HKCU\Software\Valve\Steam\Apps\209000
 Default=False
 FileKey1=%LocalAppData%\VirtualStore\Program Files*\Steam\steamapps\common\Batman Arkham Origins\SinglePlayer\BMGame\Logs|*.*
-FileKey2=%ProgramFiles%\Steam\steamapps\common\Batman Arkham Origins\SinglePlayer\BMGame\Logs|*.*
+FileKey2=%ProgramFiles%\Steam\steamapps\common\Batman Arkham Origins\SinglePlayer\BMGame\Logs
```
Comment (ROCKNROLLKID) [inline Winapp2.ini]: Maybe you did this by mistake?

File: Winapp2.ini [hunk 5]
```diff
@@ -3254,38 +3133,24 @@ Section=Games
...
-FileKey5=%LocalAppData%\VirtualStore\ProgramData\Battle.net\Client\Blizzard Launcher.*\Logs|*.*|RECURSE
-FileKey6=%LocalAppData%\VirtualStore\ProgramData\Battle.net\Client\Logs|*.*|RECURSE
+FileKey2=%CommonAppData%\Battle.net\*\Logs|*.*|RECURSE
+FileKey3=%CommonAppData%\Battle.net\Agent\Agent.*\Logs|*.*|RECURSE
+FileKey4=%CommonAppData%\Battle.net\Client\Blizzard Launcher.*\Logs|*.*|RECURSE
+FileKey5=%CommonAppData%\Battle.net\Setup\*\Logs|*.*
+FileKey6=%CommonAppData%\Battle.net\Telemetry|*.*
+FileKey7=%CommonAppData%\Blizzard Entertainment\Battle.net\Cache|*.*|REMOVESELF
+FileKey8=%LocalAppData%\Battle.net\BrowserCache|*.*|RECURSE
+FileKey9=%LocalAppData%\Battle.net\Cache|*.*|RECURSE
```
Comment (ROCKNROLLKID) [inline Winapp2.ini]: Please separate cache files in this entry, too.

File: Winapp2.ini [hunk 6]
```diff
@@ -3758,20 +3621,20 @@ Default=False
...
-FileKey1=%Documents%\My Games\Borderlands*\WillowGame\Logs|*.log
-
 [Borderlands 2*]
 Section=Games
 Detect=HKCU\Software\Valve\Steam\Apps\49520
 DetectFile=%Documents%\My Games\Borderlands 2
 Default=False
 FileKey1=%Documents%\My Games\Borderlands 2\WillowGame\Logs|*.*
 
+[Borderlands*]
```
Comment (ROCKNROLLKID) [inline Winapp2.ini]: Maybe we could combine Borderlands 1 and 2 into 1 entry and call it Borderlands 1/2*?

File: Winapp2.ini [hunk 7]
```diff
@@ -4245,33 +4096,24 @@ FileKey1=%AppData%\ACFEExamPrep|*.log
...
 DetectFile=%Documents%\Chameleon Files
 Default=False
 FileKey1=%Documents%\Chameleon Files\Cache|*.*|RECURSE
-
-[Chameleon Log*]
-LangSecRef=3024
-DetectFile=%Documents%\Chameleon Files
-Default=False
-FileKey1=%Documents%\Chameleon Files\Log|*.*|RECURSE
+FileKey2=%Documents%\Chameleon Files\Log|*.*|RECURSE
```
Comment (ROCKNROLLKID) [inline Winapp2.ini]: This cache should be separate, too.

File: Winapp2.ini [hunk 8]
```diff
@@ -253,7 +253,7 @@ Detect1=HKCU\Software\Vivaldi
 Default=False
 FileKey1=%LocalAppData%\Google\Chrome*\Application\*\Installer|*.*|RECURSE
 FileKey2=%LocalAppData%\Google\Update\*\*\Installer|*.*|RECURSE
-FileKey3=%ProgramFiles%|GUT*tmp
+FileKey3=%ProgramFiles%\|GUT*tmp
```
[the remark this answers is not in the document]
Comment (MoscaDotTo) [reply Winapp2.ini]: They're not necessary for functionality, but there was an existing rule in winappdebug that complained about them being missing so I added them. They could just as easily be trimmed and the rule removed from winappdebug

File: Winapp2.ini (hunk 1 above)
Comment (MoscaDotTo) [reply Winapp2.ini]: Yes. They didn't used to but I tested with a recent CCleaner and they seemed to be, even with nested wildcards

File: Winapp2.ini [hunk 9]
```diff
@@ -1497,7 +1478,7 @@ FileKey2=%AppData%\Adobe\Elements Organizer\*\Organizer|*.txt;status.dat
 FileKey3=%AppData%\Adobe\Elements Smart Tag Agent\*\Logs|*.log
 FileKey4=%CommonAppData%\Adobe\Elements Organizer\Catalogs\My Catalog\Watch Folder|*.txt;*.xml
 FileKey5=%CommonAppData%\Adobe\Elements Organizer\Catalogs\My Catalog|face.thumb.9.cache;thumb.5.cache
-FileKey6=%CommonAppData%\Adobe\Elements Organizer\Catalogs\My Catalog|WaldoData\waldo.cache
+FileKey6=%CommonAppData%\Adobe\Elements Organizer\Catalogs\My Catalog\WaldoData\waldo.cache
```
[the remark this answers is not in the document]
Comment (MoscaDotTo) [reply Winapp2.ini]: I saw this one and was unsure, since waldo.cache follows a slash instead of the pipe

**Comment by MoscaDotTo** [comment]
Commited another update that should address many alphabetization issues. Split back out some cache entries into their own entries once again, but have not yet completed that.

**Comment by MoscaDotTo** [comment]
I pulled a bunch of the submissions from the Piriform Forums today and am working on adding them to this refactor. Some will not be included and should be revised based on the refactor before resubmission. I will resubmit a series of updates in a piecewise fashion such that (hopefully) they can be reviewed on Git (as the alphabetization touches many thousands of lines) soon.

**Comment by MoscaDotTo** [comment]
Closing this. Please see #138

**Pull request closed without merging**